File: dart/lib/portfolio.dart
import 'package:money/money.dart';

import 'bank.dart';

class Portfolio {
  List<Money> moneys = [];

  void add(Money money) {
    moneys.add(money);
  }

  Money evaluate(Bank bank, String currency) {
    List<String> failures = [];
    var total = moneys.map((m) {
      try {
        var result = bank.convert(m, currency);
        return result?.amount;
      } catch (e) {
        failures.add('${m.currency}->$currency');
      }
    }).reduce(
      (total, amount) {
        final tempTotal = total ?? 0;
        final tempAmount = amount ?? 0;
        return tempTotal + tempAmount;
      },
    );

    if (failures.isEmpty) {
      return Money(total!, currency);
    }

    throw UnsupportedError('Missing exchange rate(s):[${failures.join(',')}]');
  }
}

File: python/portfolio.py
import functools
import operator

from money import Money


class Portfolio:
    def __init__(self) -> None:
        self.moneys = []

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ke:
                failures.append(ke)

        if len(failures) == 0:
            return Money(total, currency)

        failureMessage = ",".join(str(f) for f in failures)
        raise Exception("Missing exchange rate(s):["+failureMessage+"]")
